src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs

```diff
@@ -0,0 +1,309 @@
+
+        /// <summary>
+        /// Tags are string-string key-value pairs that represent information that will
+        /// be logged along with the Activity to the logging system.   This information
+        /// however is NOT passed on to the children of this activity.  (see Baggage)
+        /// </summary>
+        public IEnumerable<KeyValuePair<string, string>> Tags
+        {
+            get
+            {
+                if (ParentId != null)
+                    yield return new KeyValuePair<string, string>("ParentId", ParentId);
+                for (var tags = _tags; tags != null; tags = tags.Next)
+                    yield return tags.keyValue;
+            }
+        }
+
+        /// <summary>
+        /// Tags are string-string key-value pairs that represent information that will
+        /// be passed along to children of this activity.   Baggage is serialized 
+        /// when requests leave the process (along with the ID).   Typically Baggage is
+        /// used to do fine-grained control over logging of the activty and any children.  
+        /// In general, if you are not using the data at runtime, you should be using Tags 
+        /// instead. 
+        /// </summary> 
+        public IEnumerable<KeyValuePair<string, string>> Baggage
+        {
+            get
+            {
+                for (var activity = this; activity != null; activity = activity.Parent)
+                    for (var baggage = activity._baggage; baggage != null; baggage = baggage.Next)
+                        yield return baggage.keyValue;
+            }
+        }
+
+        /// <summary>
+        /// Returns the value of the key-value pair added to the activity with 'WithBaggage'.
+        /// Returns null if that key does not exist.  
+        /// </summary>
+        public string GetBaggageItem(string key)
+        {
+            foreach (var keyValue in Baggage)
+                if (key == keyValue.Key)
+                    return keyValue.Value;
+            return null;
+        }
+
+        /* Constructors  Builder methods */
+
+        /// <summary>
+        /// Note that Activity has a 'builder' pattern, where you call the constructor, a number of 'With*' APIs and then
+        /// call 'Activity.Start' to build the activity.   You MUST call Start before using it 
+        /// </summary>
+        /// <param name="operationName">Operations name <see cref="OperationName"/></param>
+        public Activity(string operationName)
+        {
+            OperationName = operationName;
+        }
+
+        /// <summary>
+        /// Update the Activity to have a tag with an additional 'key' and value 'value'.
+        /// This shows up in the 'Tags' eumeration.   It is meant for information that
+        /// is useful to log but not needed for runtime control (for the latter, use Baggage)
+        /// </summary>
+        /// <returns>'this' for convinient chaining</returns>
+        public Activity AddTag(string key, string value)
```

Comment:
Non string types are very important for tags. You want to have at least bool (e.g error state), int (e.g. Http Status code) and double (e.g sampling rate).
Requiring users to cast to string is a problem because double.ToString will use the current culture instead of InvariantCulture.
Just switching to string/object is problematic as well because in that case, people will pass complex objects.

IMO the best way is to have overloads for all accepted types (string,bool,int,double). 
Internally there should either be a string/object storage or the overloads should cast everything to string. Don't know what's better in regards of boxing etc?!
Also, if you use string/string internally, it might be problematic for the Tags-getter because users won't get the same type back as they added.